altair/datasets/_readers.py

```diff
@@ -0,0 +1,569 @@
+        fp = Path(name)
+        constraints["dataset_name"] = fp.stem
+        constraints["suffix"] = fp.suffix
+        return constraints
+    elif suffix is not None:
+        if not is_ext_read(suffix):
+            msg = (
+                f"Expected 'suffix' to be one of {EXTENSION_SUFFIXES!r},\n"
+                f"but got: {suffix!r}"
+            )
+            raise TypeError(msg)
+        else:
+            constraints["suffix"] = suffix
+    constraints["dataset_name"] = name
+    return constraints
+
+
+def validate_suffix(source: StrPath, guard: Callable[..., TypeIs[_T]], /) -> _T:
+    suffix: Any = Path(source).suffix
+    if guard(suffix):
+        return suffix
+    else:
+        msg = f"Unexpected file extension {suffix!r}, from:\n{source}"
+        raise TypeError(msg)
+
+
+def is_ext_scan(suffix: Any) -> TypeIs[_ExtensionScan]:
+    return suffix == ".parquet"
+
+
+def is_available(pkg_names: str | Iterable[str], *more_pkg_names: str) -> bool:
+    pkgs_names = pkg_names if not isinstance(pkg_names, str) else (pkg_names,)
+    names = chain(pkgs_names, more_pkg_names)
+    return all(find_spec(name) is not None for name in names)
+
+
+def infer_backend(
+    *, priority: Sequence[_Backend] = ("polars", "pandas[pyarrow]", "pandas", "pyarrow")
+) -> _Reader[Any, Any]:
+    """
+    Return the first available reader in order of `priority`.
+
+    Notes
+    -----
+    - ``"polars"``: can natively load every dataset (including ``(Geo|Topo)JSON``)
+    - ``"pandas[pyarrow]"``: can load *most* datasets, guarantees ``.parquet`` support
+    - ``"pandas"``: supports ``.parquet``, if `fastparquet`_ is installed
+    - ``"pyarrow"``: least reliable
+
+    .. _fastparquet:
+        https://github.com/dask/fastparquet
+
+    """
+    it = (backend(name) for name in priority if is_available(_requirements(name)))
+    if reader := next(it, None):
+        return reader
+    msg = f"Found no supported backend, searched:\n" f"{priority!r}"
+    raise NotImplementedError(msg)
+
+
+@overload
+def backend(name: _PolarsAny, /) -> _Reader[pl.DataFrame, pl.LazyFrame]: ...
+
+
+@overload
+def backend(name: _PandasAny, /) -> _Reader[pd.DataFrame, pd.DataFrame]: ...
+
+
+@overload
+def backend(name: _PyArrow, /) -> _Reader[pa.Table, pa.Table]: ...
+
+
+def backend(name: _Backend, /) -> _Reader[Any, Any]:
+    """Reader initialization dispatcher."""
+    if name == "polars":
+        return _PolarsReader(name)
+    elif name == "polars[pyarrow]":
+        return _PolarsPyArrowReader(name)
+    elif name == "pandas[pyarrow]":
+        return _PandasPyArrowReader(name)
+    elif name == "pandas":
+        return _PandasReader(name)
+    elif name == "pyarrow":
+        return _PyArrowReader(name)
+    elif name in {"ibis", "cudf", "dask", "modin"}:
+        msg = "Supported by ``narwhals``, not investigated yet"
+        raise NotImplementedError(msg)
+    else:
+        msg = f"Unknown backend {name!r}"
+        raise TypeError(msg)
```

Comment:
https://github.com/vega/altair/pull/3631#discussion_r1846878804

> In OP you mention that 5 backends have been implemented, but I cannot see which ones.

@mattijn these were the backends I was refering to - which correspond to the strings mentioned in `Loader.with_backend`:

https://github.com/vega/altair/blob/7ddb2a8c1e8ec6477cfc646c385e0b168f2fd330/altair/datasets/__init__.py#L66-L90

Reply:
> Thanks! My answer remain the same. Spatial support for (other) backends can also be included in follow up PRs.

Apologies @mattijn I haven't read through the whole of (https://github.com/vega/altair/pull/3631#discussion_r1846878804) - just saw one part I could help with. 

Working on feedback for (https://github.com/vega/altair/pull/3631#issuecomment-2480853741) currently - but will loop back after 😅

Reply:
Thanks for thinking this through. For what I saw, I didn't think the loader objects were stateful, so I was picturing there being a global object at the module level, but if that's no the case I wouldn't want to consider this approach. Also, extensibility outside of the altair package is potentially a good reason to avoid this approach.

Perhaps we could keep the current design and add a convenience `@classmethod` to `Loader` that would would perform the load in one step (creating a temp Loader instance internally).

```
import altair as alt
from altair.datasets import Loader
cars = Loader.load("cars", backend="polars")
```
Or perhaps `Loader` is callable
```
import altair as alt
from altair.datasets import Loader
cars = Loader("cars", backend="polars")
```

Reply:
> Thanks for thinking this through. For what I saw, I didn't think the loader objects were stateful, so I was picturing there being a global object at the module level, but if that's no the case I wouldn't want to consider this approach. 

No worries @jonmmease, it was an interesting idea and I had fun looking into it 😄.
While I was working on the whole `.with_backend()` approach, I did have [`ibis.set_backend`](https://ibis-project.org/how-to/configure/basics#default-backend) in-mind - which does work like that.

The issue I had was that AFAIK you couldn't "*reset*" the type parameters on a global object - so the IDE support wouldn't translate.

> Also, extensibility outside of the altair package is potentially a good reason to avoid this approach.

Yeah I had imagined having multiple `Loader.from_...` constructors for that reason.
Almost immediately after asking for feedback 🤦‍♂️ -  I thought one of these might be better:

```py
# Loader.with_backend()
Loader.from_backend()
Loader.from_backend_name()
```

Then for cases where someone wants to extend, they'd use a different one to parameterize these:

<details><summary><code>_readers._Reader</code></summary>
<p>

https://github.com/vega/altair/blob/7ddb2a8c1e8ec6477cfc646c385e0b168f2fd330/altair/datasets/_readers.py#L99-L120

</p>
</details>

```py
Loader.from_[placeholder_name](
    read_fn: dict[Extension, Callable[..., IntoDataFrameT]],
    scan_fn: dict[_ExtensionScan, Callable[..., IntoFrameT]],
    name: LiteralString,
)
```

Or just expose `_Reader` directly:

```py
Loader.from_reader(reader: Reader)
```

---


> Perhaps we could keep the current design and add a convenience `@classmethod` to `Loader` that would would perform the load in one step (creating a temp Loader instance internally).
> 
> ```
> import altair as alt
> from altair.datasets import Loader
> cars = Loader.load("cars", backend="polars")
> ```
> 
> Or perhaps `Loader` is callable
> 
> ```
> import altair as alt
> from altair.datasets import Loader
> cars = Loader("cars", backend="polars")
> ```

Yeah these could work and wouldn't require much to change.
The core bit we need is bringing the `backend` string into scope for an IDE to see

Would this be replacing or in addition to?:
- [feat(DRAFT): Adds infer-based `altair.datasets.load`](https://github.com/vega/altair/pull/3631/commits/7ddb2a8c1e8ec6477cfc646c385e0b168f2fd330)

Reply:
I think these are the same idea, so one or the other.  This is really great work, and I don't want to make it any more complicated for you.  My high level request is that something of this form works:

```
from altair.datasets import {something}
cars = {something that uses the name "cars" and includes an optional backend="polars" option}
```
where we feel good about showing these two lines to beginners (in basically every gallery example).
Thanks for iterating!

Reply:
@jonmmease thought this would be helpful to show alongside the commit

- [feat(DRAFT): Add optional `backend` parameter for `load(...)`](https://github.com/vega/altair/pull/3631/commits/7b3a89e5b5374eb391b7ae73ace219327069f979)

So you can see the additional `@overload`(s) are being used to support the inference here

![image](https://github.com/user-attachments/assets/bfbd2a8f-9754-4691-90bf-4191be408481)